Next iteration including new endpoints and validation

mscappini opened the pull request.

Commit:
Created 3 new endpoints: strikes, expiries, and options. Each has query parameter requirements to be met before querying and serving JSON data as results.
Now using momentjs to deal with JavaScript dates/times.

aisthesis commented on lib/app.js:
```diff
@@ -1,7 +1,7 @@
   'use strict';
 
-  process.env.NODE_ENV = process.env.NODE_ENV || 'dev';
+  process.env.NODE_ENV = process.env.NODE_ENV || 'development';
```
Are you sure you need this? I noticed when I changed the value to 'production' on my VMI, all of a sudden dev dependencies weren't pulled.  `npm` will default to 'development', so if you're not actually reading it, I don't think you need it.

mscappini replied:
Well, by default it's undefined. Under scenarios where Node.js sees that it's undefined, it assumes development. I suppose our code could work the same way, but I was putting development here to be more specific.

aisthesis commented on package.json:
```diff
@@ -1,39 +1,40 @@
+  },
+  "dependencies": {
+    "express": "~4.12.3",
+    "moment": "^2.10.2",
```
use tilde

aisthesis commented:
Include in the README under testing the code conformity tests, as in https://github.com/optbot/template#code-conformity

aisthesis commented:
When I try to start it, I'm getting:

```
$ npm start

> @optbot/restapi@0.0.2 start /mnt/disk1/home/marshall/workspace/restapi
> node lib/app.js

[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
[Error: Module did not self-register.]
js-bson: Failed to load c++ bson extension, using pure JS version
Express server started at http://:::8081
```

And since the process is blocking and seems to be running despite the errors, I tried hitting it through the browser with `http://192.168.56.101:8081/api/?equity=SCTY&option=call&strike=30&expiry=2015-4-9`. But I'm getting the simple 'Not found' page (unformatted, so it's picking up that from the `restapi` code).

mscappini commented:
Oops, forgot to update the readme. I added new endpoints to be more specific. Try `/api/options/?...` to test.

Those warnings are related to the Mongo install not finding python. I've only ever had the issue on Windows, never seen it on Mac or Ubuntu.

aisthesis commented:
now i get `{"options":[]}`

aisthesis commented:
Those _errors_ really need to be gone before we check it in, too. We can't release code known to generate errors, and warnings should only happen under special circumstances, certainly not under `npm start`

aisthesis commented on lib/routes/api.js:
```diff
@@ -1,80 +1,129 @@
 var express = require('express');
 var router = express.Router();
 
 var MongoClient = require('mongodb').MongoClient;
```
The instance of the MongoClient class should be camel case.

mscappini commented:
It's actually Mongo that gives the errors. Like I said, I've never seen that happen on anything other than Windows so I just assumed it was a Windows issue. Apparently not.

Plus the app will still run with the errors. That's why I consider them warnings. But I guess they are technically still errors...

mscappini commented:
Regarding the empty options, is your database still called `options`? It's looking for a database called `optionsMkt` but will return no data if the database is named incorrectly. This is how Mongo works out of the box, we would have to specifically check for the database and then throw an exception if it doesn't exist if that's the behavior we're looking for.

Commit:
Added code conformity to README.
Removed unnecessary NODE_ENV set.
Camel cased Mongo client instance.
Updated package.json with proper semver.

Commit:
Removed the 'test' folder from code conformity.

aisthesis commented:
yes, and it will continue to be called options for a bit at least. Could you call out in the README how to configure it appropriately?

aisthesis commented:
Errors are gone after re-installing the node_modules folder.

aisthesis commented:
One other thing: Could you make the 'start' instruction in the README be `sudo npm start`? I realize that it doesn't matter at the moment for this service, but `sudo` should at least not break it.

The reason I'd like it as that each service that stays up should run as its own user and group, so if you or I start it as ourselves rather than root, we actually shouldn't be able to. Root will have the ability to act on behalf of any user and group, so also the one running the service. I've implemented this for `eqpuller` following the way `mongo` sets up its running process using the user and group `mongodb`.

Also the `npm config` commands given as examples won't work without `sudo` because global configurations are stored under root control. So, could you add `sudo` to those example commands as well.

Commit:
Included instructions in the README how to redefine the MongoDB connection string.

aisthesis commented on README.md:
```diff
@@ -33,11 +33,22 @@ Testing
 
+If no data is being returned, it may be that the database connection string is improperly configured. As outlined in the **Connects To** section, this project currently requires a database called `optionsMkt` with a collection named `quotes`. To reconfigure this, you can issue the following command:
+
+	$ npm config set @optbot/restapi:db:optionsMkt "mongodb://<your_mongo_host>:<your_mongo_port>/<your_database_name>"
```
missing `sudo` and `--global`

aisthesis commented on README.md:
```diff
@@ -33,11 +33,22 @@ Testing
+
+As an example, using the default settings would be
+
+	$ npm config set @optbot/restapi:db:optionsMkt "mongodb://localhost:27017/optionsMkt"
```
missing `sudo` and `--global`

mscappini commented:
I would prefer to not have `sudo` in the instructions because I am trying to keep the instructions as platform-agnostic as possible. So far, these instructions will work on Windows as well. I'm sure if you're using some form of Linux, you know what `sudo` does and will use it where appropriate.

In a related note: I was reading that `sudo` is [not recommended](https://github.com/npm/npm/tree/99f804f43327c49ce045ae2c105995636c847145#permissions) for `npm start`.

Commit:
Added --global to npm config.

aisthesis commented:
How will we be able to run the process as user `restapi` and group `restapi` otherwise?

aisthesis commented:
I ran `sudo npm config set @optbot/restapi:db:optionsMkt "mongodb://localhost:27017/options" --global` but am still getting no data.

aisthesis commented:
The reason I prefer `sudo` in the README is because, while it is ok to be messing around with instructions in development, installing and running on production should be pure copy-paste. And the production server is Linux. When I install there, I want to know exactly what to do without having to think about it. On Windows it is easily recognizable that sudo won't work, you can leave it out on your copy-paste, and it simply won't do anything if you forget. On Linux, it is possible that leaving it out will do a bunch of stuff that you don't want to the user that is running `npm start` because the system _will_ understand what it means to run it without `sudo`.

I wasn't able to find anything about not running `npm start` as root in the link you gave. All I saw was that it is ok to run some commands as root.

mscappini commented:
But you could make the same argument for *nix. You should know to run as root, so include `sudo`. Basically, do what is appropriate for your OS. I don't believe README instructions should be a copy/paste job. They are guidelines, and things should work out of the box, but setting up without even thinking just seems like a mistake to begin with.

mscappini commented:
But I also have another thought on that. It's been to my understanding that you should never run a webserver as root for security reasons. Wouldn't using `sudo` run this webserver as root?

aisthesis commented:
The instructions should apply to the production servers and should be formed in such a way to later be automated. The reason for including `sudo` now is to make sure there are no unanticipated side effects that don't show actual errors.

If, on production, I happen to type `npm config set foo bar` and then `npm start` with no `sudo`, then the process will run for me. But then the configuration won't apply if you go to the same server because the config settings will be saved in files specific to my user.  Then perhaps you'll go and do your config in the same way, and soon we have the same service on the same server running in different ways depending on who starts it. The problem on Linux is that it's non-trivial to know when to use it and when not (in contrast to the 'nix' command, where the answer is always 'not').

aisthesis commented:
As to your point of running it as root, the point is actually to avoid that. The difficulties I had with `eqpuller` involved creating a specific `eqpuller` user and a specific `eqpuller` group that can write to the `eqpuller` logs and run the `eqpuller` process but can do nothing else. That's how we should configure these systems in my opinion. For now, if I don't use sudo, I'll be starting it on production as 'marshall', which is also wrong. What `sudo` does is allow you freedom to specify what user is running the process when you start it. For example: https://github.com/optbot/eqpuller/blob/marshall/lib/start.js#L24

mscappini commented:
I understand the desire to want `sudo` for that reason. But when I setup an IIS server, I always run it with an app pool using a particular user. This is knowledge that I know and understand as someone who sets up websites on Windows. The same knowledge should still apply to you as someone who sets up a server running on any OS you desire to run this instance on--including Linux. In which case, if you are doing this, you would know to use `sudo` where appropriate.

mscappini commented:
I understand using `sudo` to specify the user. You should do that, and also know to use `sudo` to run it as that user.

mscappini commented:
The point about automation I believe to actually be in favor of not putting `sudo` in the instructions. If it's automated, all you'll have to do is remember to include it once. Obviously this will be tested thoroughly before we continue to automate everything, so having to write it into the script only once shouldn't be a hassle.

mscappini commented:
And wouldn't using `sudo` to run as a user negate the copy/paste anyway? You have to specify the user to run as.

aisthesis commented:
No, the creation of the user and the user with which to run the code all happen in code. [This](https://github.com/mongodb/mongo/blob/7423c26ab95464a9b8f620ab4a11c297dc0621d5/debian/mongodb-org-server.postinst#L24) is how mongo accomplishes the user creation, for example, which is why you have to run the installation as `sudo`. We don't have anything nearly as elaborate as mongo, but to do it right, we do have to create a new user with permissions to run the appropriate executables and with permissions to write to the appropriate log directory.

mscappini commented:
I see, can you point me to where this all takes place so I can better understand it?

aisthesis commented:
https://github.com/optbot/eqpuller/blob/marshall/lib/setup.js

I wouldn't worry about it too much for now. I know you prefer developing on Windows, and this is a Linux thing. I'm happy to check out the code and make the necessary changes when the time comes. For now, since it's only running temporarily, it's no big deal anyway, and I'm not worried about who owns the process.

Once the service is publicly exposed and it does become an issue, It might be possible to set a flag on `npm start` that bypasses platform-specific code so you can continue to develop on Windows if you like. I'd be for that as long as the branching stays fairly tame. The commands for user creation and managing processes under a certain user are clearly not going to work on both platforms, but we could have something like `npm start --win` that runs it under Windows, bypassing the user-creation steps and similar.

Again, this isn't meant for this iteration but for a few cycles down the road. For this iteration, I just want to know what I'm doing when I install in production without having a separate text file of my own to record the commands.

mscappini commented:
Sounds good. I'll have a look when I get back to a computer.

I fixed the query returning no results. As usual, it was a timezone thing.

Commit:
Fixed querying of expiry and date TZ issues.

aisthesis commented:
It's working now. My db on VirtualBox just isn't populated with the expiry date of the example.

mscappini commented:
Excellent. I wish I had an example that would always certainly work.

Commit:
Added sudo to npm start and npm config to alleviate permissions issues.

Commit:
MOAR SUDO

aisthesis commented:
Ship it!
